flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/SequentialChannelStateReaderImpl.java

```diff
@@ -0,0 +1,263 @@
+		return handle -> Preconditions.checkState(seen.add(handle.getInfo()), "duplicate channel info: %s");
+	}
+
+	private static <Info, Handle extends AbstractChannelStateHandle<Info>> List<Tuple2<Long, Info>> extractOffsetsSorted(List<Handle> channelStateHandles) {
+		return channelStateHandles
+			.stream()
+			.flatMap(SequentialChannelStateReaderImpl::extractOffsets)
+			.sorted(comparingLong(offsetAndInfo -> offsetAndInfo.f0))
+			.collect(toList());
+	}
+
+	private static  <Info, Handle extends AbstractChannelStateHandle<Info>> Stream<Tuple2<Long, Info>> extractOffsets(Handle handle) {
+		return handle.getOffsets().stream().map(offset -> Tuple2.of(offset, handle.getInfo()));
+	}
+
+	@Override
+	public void close() throws Exception {
+	}
+
+}
+
+interface RecoveredChannelStateHandler<Info, Context> extends AutoCloseable {
+	Tuple2<ChannelStateByteBuffer, Context> getBuffer(Info info) throws IOException;
+
+	void recover(Info info, Context context) throws IOException;
+}
+
+class InputChannelRecoveredStateHandler implements RecoveredChannelStateHandler<InputChannelInfo, Buffer> {
+	private final InputGate[] inputGates;
+
+	InputChannelRecoveredStateHandler(InputGate[] inputGates) {
+		this.inputGates = inputGates;
+	}
+
+	@Override
+	public Tuple2<ChannelStateByteBuffer, Buffer> getBuffer(InputChannelInfo channelInfo) throws IOException {
+		RecoveredInputChannel channel = getChannel(channelInfo);
+		Buffer buffer;
+		try {
+			buffer = channel.getBuffer();
+			return Tuple2.of(wrap(buffer), buffer);
+		} catch (InterruptedException e) {
+			throw new IOException(e);
+		}
+	}
+
+	@Override
+	public void recover(InputChannelInfo channelInfo, Buffer buffer) {
+		if (buffer.readableBytes() > 0) {
+			getChannel(channelInfo).onRecoveredStateBuffer(buffer);
+		} else {
+			buffer.recycleBuffer();
+		}
+	}
+
+	@Override
+	public void close() throws IOException {
+		// note that we need to finish all RecoveredInputChannels, not just those with state
+		for (final InputGate inputGate : inputGates) {
+			inputGate.finishReadRecoveredState();
+		}
+	}
+
+	private RecoveredInputChannel getChannel(InputChannelInfo info) {
+		return (RecoveredInputChannel) inputGates[info.getGateIdx()].getChannel(info.getInputChannelIdx());
+	}
+}
+
+class ResultSubpartitionRecoveredStateHandler implements RecoveredChannelStateHandler<ResultSubpartitionInfo, Tuple2<BufferBuilder, BufferConsumer>> {
+
+	private final ResultPartitionWriter[] writers;
+
+	ResultSubpartitionRecoveredStateHandler(ResultPartitionWriter[] writers) {
+		this.writers = writers;
+	}
+
+	@Override
+	public Tuple2<ChannelStateByteBuffer, Tuple2<BufferBuilder, BufferConsumer>> getBuffer(ResultSubpartitionInfo subpartitionInfo) throws IOException {
+		BufferBuilder bufferBuilder;
+		try {
+			bufferBuilder = getWriter(subpartitionInfo).getBufferBuilder(subpartitionInfo.getSubPartitionIdx());
+		} catch (InterruptedException e) {
+			throw new IOException(e);
```

Comment:
`InterruptedException` is a checked exception so I'd had to declare it if not wrapping. 
But it's not appropriate for this layer (IO).

`Thread.currentThread().interrupt();` only sets the status, so the caller can skip interrupt status check.

Reply:
`InterruptedException` quite often should be handled differently and it's good to know that the `InterruptedException` has happened.

For example, when caller received `IOException`, it might decide to retry or just keep going if that was not a critical part. `InterruptedException` should on the other hand shut down the code semi gracefully and as quickly as possible.

Here I think adding `throws InterruptedException` would be much better. But I didn't look how deep in a call stack are we and how difficult would it be to propagate it.

re `Thread.currentThread().interrupt();`, the crowd wisdom is to always set it when wrapping/hiding/rethrowing `InterruptedException`. I think we are using interrupted flag in only a couple of places, but it's better to keep it clean, just in case we would had to use it some important place and then we would discover we need to fix `interrupted` flag everywhere to make it work.

Reply:
You're right about `currentThread().interrupt()` and that handling of `IOException` likely differs from that of `InterruptedException`. 
But for the latter, I think it's better to wrap it into `RuntimeException`.

Reply:
Personally I don't like declared exceptions, but even in that case special treatment of the `InterruptedException` means  it makes sense to declare it explicitly. But the issue is that in Flink trend is to have explicit exceptions handling/declarations. Here it looks like the `InterruptedException` is already present on the call site (`StreamTask#readRecoveredChannelState`), so it shouldn't be an issue to conform with this.

Reply:
Removed `catch InterruptedException` and added to the method signature after not converging in an offline discussion.